tests/test_media_multipart.py

```diff
@@ -0,0 +1,495 @@
+        'multipart/form-data; boundary=' + boundary,
+        len(example))
+
+    for part in form:
+        output = io.BytesIO()
+        part.stream.pipe(output)
+        assert isinstance(output.getvalue(), bytes)
+
+
+@pytest.mark.parametrize('buffer_size,chunk_size', list(itertools.product(
+    (
+        32,
+        64,
+        128,
+        256,
+    ),
+    (
+        7,
+        8,
+        9,
+        10,
+        32,
+        64,
+        128,
+        256,
+    ),
+)))
+def test_parsing_correctness(buffer_size, chunk_size):
+    example = EXAMPLES['boundary']
+    handler = media.MultipartFormHandler()
+    stream = BufferedStream(io.BytesIO(example).read, len(example),
+                            buffer_size)
+    form = handler.deserialize(
+        stream, 'multipart/form-data; boundary=boundary', len(example))
+
+    for part in form:
+        if part.name in ('lorem1', 'lorem2'):
+            part_stream = part.stream
+            result = []
+            while True:
+                chunk = part_stream.read(chunk_size)
+                if not chunk:
+                    break
+                result.append(chunk)
+
+            assert b''.join(result) == LOREM_IPSUM
+
+
+def test_missing_boundary():
+    handler = media.MultipartFormHandler()
+
+    with pytest.raises(falcon.HTTPInvalidHeader):
+        handler.deserialize(io.BytesIO(), 'multipart/form-data', 0)
+
+    with pytest.raises(falcon.HTTPInvalidHeader):
+        handler.deserialize(io.BytesIO(), 'multipart/form-data; boundary=', 0)
+
+    overlong = '-' * 71
+    content_type = 'multipart/form-data; boundary=' + overlong
+    with pytest.raises(falcon.HTTPInvalidHeader):
+        handler.deserialize(io.BytesIO(), content_type, 0)
+
+
+def test_empty_input():
+    handler = media.MultipartFormHandler()
+    form = handler.deserialize(
+        io.BytesIO(), 'multipart/form-data; boundary=404', 0)
+    with pytest.raises(falcon.HTTPBadRequest):
+        for part in form:
+            pass
+
+
+def test_serialize():
+    handler = media.MultipartFormHandler()
+    with pytest.raises(NotImplementedError):
+        handler.serialize({'key': 'value'}, 'multipart/form-data')
+
+
+def test_content_transfer_encoding_header():
+    data = (
+        b'--BOUNDARY\r\n'
+        b'Content-Disposition: form-data; name="file"; filename="bytes"\r\n'
+        b'Content-Transfer-Encoding: Base64'
+        b'Content-Type: application/x-falcon\r\n\r\n'
+        b'UGVyZWdyaW5lIEZhbGNvbiADLgA='
+        b'\r\n'
+        b'--BOUNDARY\r\n'
+        b'Content-Disposition: form-data; name="empty"\r\n'
+        b'Content-Type: text/plain\r\n\r\n'
+        b'\r\n'
+        b'--BOUNDARY--\r\n'
+    )
+
+    handler = media.MultipartFormHandler()
+    form = handler.deserialize(
+        io.BytesIO(data), 'multipart/form-data; boundary=BOUNDARY', len(data))
+    with pytest.raises(falcon.HTTPBadRequest):
+        for part in form:
+            pass
+
+
+def test_unsupported_charset():
```

Comment:
Could you also add tests for mismatched charsets (e.g., charset says it is one thing, but the actual data fails to decode)?